processor/src/advice/source.rs

```diff
@@ -0,0 +1,30 @@
+use super::{Felt, Word};
+
+// ADVICE SOURCE
+// ================================================================================================
+
+/// Placeholder for advice provider tape mutation.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum AdviceSource {
+    /// Writes a single value to the advice tape.
+    Value(Felt),
+
+    /// Fetch a keyed tape from the values map, reversing and appending it to the advice tape.
+    ///
+    /// Note: this operation shouldn't consume the map element so it can be called multiple times
+    /// for the same key.
+    ///
+    /// # Example
+    /// Given an advice tape `[a,b,c]`, and a map `x |-> [d,e,f]`, a call `write_tape_from_map(x)`
+    /// will result in `[a,b,c,f,e,d]` for the advice tape, and will preserve `x |-> [d,e,f]`.
+    ///
+    /// # Errors
+    /// Returns an error if the key was not found in the key-value map.
+    Map { key: Word },
+}
+
+impl From<Felt> for AdviceSource {
+    fn from(value: Felt) -> Self {
+        Self::Value(value)
+    }
+}
```

Comment:
I wonder if it is better to have explicit constructors (e.g., `from_value()`, `from_map()` rather than an implicit conversion.

Reply:
Implicit conversion might be too magic indeed, but on the other hand enums already have their own form of `from_*` as to create a value from an element, instead of calling `AdviceSource::from_value(foo)`, we would `AdviceSource::Value(foo)`.

`Felt` could be an exception here because this is the most atomic possible value that can go to a tape, and this can't be broken down into smaller types. If that was the case (such as in `Word`), then we could face some ambiguity. wdyt?

Reply:
Yeah, I wouldn't add an implicit conversion for `Word` as it might have multiple meanings in the future. It is possible that `Felt` might also have multiple meanings - but probably not very likely.

If you prefer to keep `From<Felt>` here - I'm fine with it.

Reply:
I also have mixed feelings to make implicit conversions so better just remove it indeed. Fixed!